### PR TITLE
Move Android interop test app to use AbstractInteropTest+Protolite

### PR DESCRIPTION
**Pull request by ericgribkoff**
This resolves https://github.com/grpc/grpc-java/issues/3968, and also gets rid of a large amount of redundant code: instead of having re-implementations of all interop tests specifically for Android, our test app now runs the shared test definitions from `AbstractInteropTest`.

The first commit makes changes to `AbstractInteropTest` to allow it to support Android and Java Proto Lite. The second commit switches the interop test to use `AbstractInteropTest`, and also cleans up/modernizes the instrumentation tests.

### CODE REVIEW

**Comment by ericgribkoff** [comment]
@dapengzhang0 friendly ping. The two commits can be reviewed separately; the second is large but just a refactor, and removes a lot of duplicate code (plus gets rid of the outdated nano usage).

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 1]
```diff
@@ -1784,7 +1802,14 @@ private static void verifyNoMoreInteractions(Object... mocks) {
     } catch (AssertionError e) {
       String msg = e.getMessage();
       if (msg.length() >= 256) {
-        throw new AssertionError(msg.substring(0, 256), e);
+        try {
+          // AssertionError(String, Throwable) only present in Android API 19+
+          throw AssertionError.class
```
Comment (dapengzhang0) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Ditto.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 2]
```diff
@@ -421,9 +424,11 @@ private StreamingOutputCallResponse toResponse() {
...
       ByteString dataBuffer = compressable ? compressableBuffer : uncompressableBuffer;
       ByteString payload = generatePayload(dataBuffer, offset, length);
-      responseBuilder.getPayloadBuilder()
-          .setType(compressable ? PayloadType.COMPRESSABLE : PayloadType.UNCOMPRESSABLE)
-          .setBody(payload);
+      responseBuilder.setPayload(
+          Payload.newBuilder()
+              .setType(compressable ? PayloadType.COMPRESSABLE : PayloadType.UNCOMPRESSABLE)
+              .setBody(payload)
+              .build());
```
Comment (dapengzhang0) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Is `.build()` needed?

File: android-interop-testing/app/proguard-rules.pro [hunk 3]
```diff
@@ -18,6 +18,8 @@
 -dontwarn sun.reflect.**
 # Ignores: can't find referenced class javax.lang.model.element.Modifier
 -dontwarn com.google.errorprone.annotations.**
+-dontwarn org.apache.commons.logging.**
 
 # Ignores: can't find referenced method from grpc-testing's compileOnly dependency on Truth
 -dontwarn io.grpc.testing.DeadlineSubject
+
```
Comment (dapengzhang0) [inline android-interop-testing/app/proguard-rules.pro]: revert newline?

File: android-interop-testing/app/proguard-rules.pro [hunk 4]
```diff
@@ -18,6 +18,8 @@
 -dontwarn sun.reflect.**
 # Ignores: can't find referenced class javax.lang.model.element.Modifier
 -dontwarn com.google.errorprone.annotations.**
+-dontwarn org.apache.commons.logging.**
```
Comment (dapengzhang0) [inline android-interop-testing/app/proguard-rules.pro]: where is apache logging from? add a comment?

File: android-interop-testing/app/src/androidTest/java/io/grpc/android/integrationtest/InteropInstrumentationTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2017 The gRPC Authors
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply android-interop-testing/app/src/androidTest/java/io/grpc/android/integrationtest/InteropInstrumentationTest.java]: Github's UI and I disagree on this being a new file - I renamed InteropTesterTest.java to make this, so I think it's ok to keep the original 2017 copyright date.

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTask.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import io.grpc.ClientInterceptor;
+import io.grpc.ManagedChannel;
+import io.grpc.testing.integration.AbstractInteropTest;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.ref.WeakReference;
+import java.util.List;
+import org.junit.AssumptionViolatedException;
+
+/** Javadoc. */
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTask.java]: Yes, every class needs Javadoc :D  Replaced with real-ish content

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 7]
```diff
@@ -301,6 +305,10 @@ public void tearDown() throws Exception {
 
   protected abstract ManagedChannel createChannel();
 
+  protected ClientInterceptor[] getAdditionalInterceptors() {
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 8]
```diff
@@ -1769,7 +1780,14 @@ private static void assumeEnoughMemory() {
     } catch (AssertionError e) {
       String msg = e.getMessage();
       if (msg.length() >= 256) {
-        throw new AssertionError(msg.substring(0, 256), e);
+        try {
+          // AssertionError(String, Throwable) only present in Android API 19+
+          throw AssertionError.class
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java (hunk 1 above)
Comment (ericgribkoff) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 2 above)
Comment (ericgribkoff) [reply interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: No, removed

File: android-interop-testing/app/proguard-rules.pro (hunk 4 above)
Comment (ericgribkoff) [reply android-interop-testing/app/proguard-rules.pro]: No longer necessary, reverted

File: android-interop-testing/app/proguard-rules.pro (hunk 3 above)
Comment (ericgribkoff) [reply android-interop-testing/app/proguard-rules.pro]: Done

**Pull request merged**